share/eds/byzantine/share_proof.go

```diff
@@ -75,21 +75,23 @@ func GetProofsForShares(
 	shares [][]byte,
 ) ([]*ShareWithProof, error) {
 	proofs := make([]*ShareWithProof, len(shares))
-	for index, share := range shares {
-		if share != nil {
-			proof := make([]cid.Cid, 0)
-			// TODO(@vgonkivs): Combine GetLeafData and GetProof in one function as the are traversing the same
-			// tree. Add options that will control what data will be fetched.
-			s, err := ipld.GetLeaf(ctx, bGetter, root, index, len(shares))
-			if err != nil {
-				return nil, err
-			}
-			proof, err = ipld.GetProof(ctx, bGetter, root, proof, index, len(shares))
-			if err != nil {
-				return nil, err
-			}
-			proofs[index] = NewShareWithProof(index, s.RawData(), proof)
+	for index, sh := range shares {
+		if sh == nil {
+			continue
+		}
+
+		proof := make([]cid.Cid, 0)
+		// TODO(@vgonkivs): Combine GetLeafData and GetProof in one function as the are traversing the same
+		// tree. Add options that will control what data will be fetched.
+		s, err := share.GetShare(ctx, bGetter, root, index, len(shares))
```

Comment:
This is where the main fix is placed. Mainly instead of trimming the namespace on the validator side, we should not send it at all and trim it on the sender(FN) side. This is what `GetShare` does. It calls the substituted `ipld.GetLeaf` and trims the namespace. `share` pkg has a bunch of proxy getters that hide leaky namespace by subslicing it. We should have done this from the beginning but somehow missed it.

@evan-forbes, @renaynay, @walldiss

Reply:
makes sense!

Reply:
@evan-forbes, I completely forgot that we need that namespace during share inclusion verification, and just using `share.GetShare` fixes the FalsePostive test, but breaks others. I tried to dynamically prepend a NID, similarly to what NMTWrapper does, but it increased hairiness, looked complex and required more params for `share.Validate`, so I reverted it to your original solution with the goal of simplifying code that it is already hard to understand